### PR TITLE
Improvements for sound FX and added ADPCM long stream support.

### PR DESCRIPTION
**Pull request by DC-SWAT**
- Improved sound FX samples loading.
- Added support for ADPCM streams.
- Added spu_memload_sq().
- Changed API for snd_pcm16_split_sq().
- Added G2 Bus locking for SQ's with SPU memory.
- Workaround for sound stream memory allocation bug.

### CODE REVIEW

File: kernel/arch/dreamcast/include/dc/sound/stream.h [hunk 1]
```diff
@@ -238,7 +238,7 @@ void snd_stream_queue_disable(snd_stream_hnd_t hnd);
 */
 void snd_stream_queue_go(snd_stream_hnd_t hnd);
 
-/** \brief  Start a stream.
+/** \brief  Start PCM 16-bit stream.
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/sound/stream.h:241]:
```suggestion
/** \brief  Start a 16-bit PCM stream.
```

I feel like this flows better...

File: kernel/arch/dreamcast/include/dc/sound/stream.h [hunk 2]
```diff
@@ -249,6 +249,17 @@ void snd_stream_queue_go(snd_stream_hnd_t hnd);
 */
 void snd_stream_start(snd_stream_hnd_t hnd, uint32 freq, int st);
 
+/** \brief  Start ADPCM 4-bit stream.
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/sound/stream.h:252]:
```suggestion
/** \brief  Start a 4-bit ADPCM stream.
```

File: kernel/arch/dreamcast/include/dc/spu.h [hunk 3]
```diff
@@ -37,6 +39,20 @@ void spu_write_wait(void);
 */
 void spu_memload(uint32 to, void *from, int length);
 
+
+/** \brief  Copy a block of data to sound RAM.
+
+    This function acts much like memcpy() but copies to the sound RAM area
+    by using SQ.
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/spu.h:46]:
```suggestion
    by using the store queues.
```

File: kernel/arch/dreamcast/include/dc/spu.h [hunk 4]
```diff
@@ -1,7 +1,8 @@
 /* KallistiOS ##version##
 
    dc/spu.h
-   Copyright (C) 2000-2001 Megan Potter
+   Copyright (C) 2000,2001 Megan Potter
```
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/spu.h:4]:
```suggestion
   Copyright (C) 2000, 2001 Megan Potter
```

File: kernel/arch/dreamcast/hardware/spu.c [hunk 5]
```diff
@@ -1,11 +1,13 @@
 /* KallistiOS ##version##
 
    spu.c
-   (c)2000-2001 Megan Potter
+   Copyright (C) 2000,2001 Megan Potter
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/spu.c:4]:
```suggestion
   Copyright (C) 2000, 2001 Megan Potter
```

File: kernel/arch/dreamcast/hardware/spu.c [hunk 6]
```diff
@@ -62,6 +64,38 @@ void spu_memload(uint32 dst, void *src_void, int length) {
     }
 }
 
+void spu_memload_sq(uint32 dst, void *src_void, int length) {
+    uint8 *src = (uint8 *)src_void;
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/spu.c:68]: Please keep variable declarations at the beginning of a scope.

```suggestion
    uint8 *src = (uint8 *)src_void;
    int aligned_len, old;
```

File: kernel/arch/dreamcast/hardware/spu.c [hunk 7]
```diff
@@ -62,6 +64,38 @@ void spu_memload(uint32 dst, void *src_void, int length) {
...
+       count to a 32-bit word count */
+    if(length & 3) {
+        length = (length + 4) & ~3;
+    }
+
+    /* Add in the SPU RAM base (cached area) */
+    dst += 0x00800000;
+
+    /* Using SQs for all that is divisible by 32 */
+    const int aligned_len = length & ~31;
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/spu.c:80]:
```suggestion
    aligned_len = length & ~31;
```

File: kernel/arch/dreamcast/hardware/spu.c [hunk 8]
```diff
@@ -62,6 +64,38 @@ void spu_memload(uint32 dst, void *src_void, int length) {
...
+    length &= 31;
+
+    int old = irq_disable();
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/spu.c:83]:
```suggestion
    old = irq_disable();
```

File: kernel/arch/dreamcast/hardware/spu.c [hunk 9]
```diff
@@ -62,6 +64,38 @@ void spu_memload(uint32 dst, void *src_void, int length) {
...
+    do { } while(*(vuint32 *)0xa05f688c & (1 << 5)) ; // FIFO_SH4
+    do { } while(*(vuint32 *)0xa05f688c & (1 << 4)) ; // FIFO_G2
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/spu.c:85]:
```suggestion
    do { } while(*(vuint32 *)0xa05f688c & (1 << 5)); /* FIFO_SH4 */
    do { } while(*(vuint32 *)0xa05f688c & (1 << 4)); /* FIFO_G2 */
```

File: kernel/arch/dreamcast/hardware/spu.c [hunk 10]
```diff
@@ -62,6 +64,38 @@ void spu_memload(uint32 dst, void *src_void, int length) {
...
+    sq_cpy((void *)dst, src, aligned_len);
+    irq_restore(old);
+
+    if(length > 0) {
+        /* Non-cached area */
+        dst += 0xa0000000;
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/spu.c:91]: This should almost certainly be an `|=` not a `+=`.

```suggestion
        /* Make sure the destination is in a non-cached area */
        dst |= 0xa0000000;
```

**Review by ljsebald — CHANGES_REQUESTED** [review]

File: kernel/arch/dreamcast/hardware/spu.c (hunk 10 above)
Comment (QuzarDC) [reply kernel/arch/dreamcast/hardware/spu.c:91]: And maybe using the nice new un-magic'd MEM_AREA_P2_BASE?

File: kernel/arch/dreamcast/hardware/spu.c [hunk 11]
```diff
@@ -1,11 +1,13 @@
...
 
    spu.c
-   (c)2000-2001 Megan Potter
+   Copyright (C) 2000,2001 Megan Potter
+   Copyright (C) 2023 Ruslan Rostovtsev
  */
 
 #include <dc/spu.h>
 #include <dc/g2bus.h>
+#include <dc/sq.h>
```
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/spu.c:10]:
```suggestion
#include <dc/sq.h>
#include <arch/memory.h>
```

File: kernel/arch/dreamcast/hardware/spu.c (hunk 10 above)
Comment (ljsebald) [reply kernel/arch/dreamcast/hardware/spu.c:91]: Good point. Updated to use the macro and another suggestion added to include the appropriate header file (even though it is indirectly included through sq.h).

**Review by ljsebald — APPROVED** [review]

**Pull request merged**